public/js/controllers/address_controller.js

```diff
@@ -80,36 +83,189 @@ function plotGraph (processedData, otherOptions) {
   )
 }
 
+function hashHighLight (matchHash, hoverOn) {
+  $('.hash').each(function () {
+    var thisHash = $(this).attr('href')
+    if (thisHash === matchHash && hoverOn) {
+      $(this).addClass('matching-hash')
+    } else {
+      $(this).removeClass('matching-hash')
+    }
+  })
+}
+
+class TurboQuery {
+  constructor (turbolinks) {
+    var ta = this
+    ta.replaceTimer = 0
+    ta.appendTimer = 0
+    ta.turbolinks = turbolinks || Turbolinks || false
+    if (!ta.turbolinks) {
+      console.error('No passed or global Turbolinks instance detected. TurboQuery requires Turbolinks.')
+      return
+    }
+    // These are timer callbacks. Bind them to the TurboQuery instance.
+    ta.replaceHistory = ta._replaceHistory.bind(ta)
+    ta.appendHistory = ta._appendHistory.bind(ta)
+    ta.url = Url(window.location.href, true)
+  }
+
+  replaceHref () {
+    // Rerouting through timer to prevent spamming.
+    // Turbolinks blocks replacement if frequency too high.
+    if (this.replaceTimer === 0) {
+      this.replaceTimer = setTimeout(this.replaceHistory, 250)
+    }
+  }
+
+  toHref () {
+    if (this.appendTimer === 0) {
+      this.appendTimer = setTimeout(this.appendHistory, 250)
+    }
+  }
+
+  _replaceHistory () {
+    // see https://github.com/turbolinks/turbolinks/issues/219. This also works:
+    // window.history.replaceState(window.history.state, this.addr, this.url.href)
+    this.turbolinks.controller.replaceHistoryWithLocationAndRestorationIdentifier(this.turbolinks.Location.wrap(this.url.href), this.turbolinks.uuid())
+    this.replaceTimer = 0
+  }
+
+  _appendHistory () {
+    // same as replaceHref, but creates a new entry in history for navigating
+    // with the browsers forward and back buttons. May still not work because of
+    // TurboLinks caching behavior, I think.
+    this.turbolinks.controller.pushHistoryWithLocationAndRestorationIdentifier(this.turbolinks.Location.wrap(this.url.href), this.turbolinks.uuid())
+    this.appendTimer = 0
+  }
+
+  replace (query) {
+    this.url.set('query', this.filteredQuery(query))
+    this.replaceHref()
+  }
+
+  to (query) {
+    this.url.set('query', this.filteredQuery(query))
+    this.toHref()
+  }
+
+  filteredQuery (query) {
+    var filtered = {}
+    Object.keys(query).forEach(function (key) {
+      var v = query[key]
+      if (typeof v === 'undefined' || v === null) return
```

Comment:
I'm a bit out of my element here, but this struck me as a place for `continue` instead of `return`.  I assume this is processing all of the query key-values in a url.  And this is catching queries with no value set?

Reply:
You actually give `forEach` a function to which each element is passed. Here, I am passing an anonymous function that is defined on the fly, so there is no loop to `continue` from. 

Even though I know that, I almost always put `continue` and encounter the error, so you're in good company.

Reply:
Ah, I see now.  Thanks.